ONSdigital/eq-author-app
# EAR 610 - Remove default "AND" from logic conditions

### What is the context of this PR?

Re-design of routing expression editor per EAR-610 to remove "AND" as the default group logical operator. Prompts the user for operator & provides validation error for this - but only if more than one expression is present.

[Prototype](https://overflow.io/s/5VB5D7H7?node=97a4f049&prototype)
[Ticket](https://collaborate2.ons.gov.uk/jira/browse/EAR-610)

### Summary of changes

- [x] Preliminary tidy up / refactor of `BinaryExpressionEditor`
- [x] Add tooltips to buttons as per AC
- [x] Fix CSS transitions
- [x] Move condition selector from title of `BinaryExpressionEditor` to new location
- [x] Add AJV validation rules for `expressionGroup.condition`
- [x] Make `expressionGroup.condition` start out `null` for routing rules & check publisher still outputs correct rules when only one expression present
- [x] Change "If" and "Else If" labels to uppercase after discussion w/ Joe
- [x] UI validation error for expression group operator ("Select AND/OR" in select & "Selection required" message) per prototype
- [x] Clear AND/OR selection when penultimate expression removed so Author "doesn't remember" - new optional callback `onExpressionDeleted` added to `BinaryExpressionEditor` to achieve this (used by routing `RuleEditor`) without disturbing existing skip logic functionality

### How to review

See the extensive AC on the [ticket](https://collaborate2.ons.gov.uk/jira/browse/EAR-610).

In addition to those:
- View survey & make sure publisher-generated routing rules still work as intended
- Check skip logic editor still works as before (the binary expression editor is shared between routing & skip)

### Screenshots
<img width="731" alt="ExprEditor1" src="https://user-images.githubusercontent.com/60441612/106958515-60cc6b00-6731-11eb-9cc7-c43f2f4e15ad.png">

-----

<img width="725" alt="Expr2" src="https://user-images.githubusercontent.com/60441612/106958521-63c75b80-6731-11eb-8e49-d65d4196a494.png">


### What to do after everything is green

1. - Bring this branch up-to-date with Origin/Master
2. - If there are a lot of commits or some are not helpful to read, squash them down
3. - Click the **Merge** button on this pull request
4. - Does this change mean the **Capability examples** questionnaire should be updated?
5. - Move the Jira ticket for this task into the next stage of the process
6. - ☕
